driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java

```diff
@@ -346,6 +371,64 @@ public boolean isClosed() {
     @Nullable
     @Override
     public <T> T sendAndReceive(final CommandMessage message, final Decoder<T> decoder, final OperationContext operationContext) {
+        try {
+            return sendAndReceiveInternal(message, decoder, operationContext);
+        } catch (MongoCommandException e) {
+            if (reauthenticationIsTriggered(e)) {
+                return reauthenticateAndRetry(()-> sendAndReceiveInternal(message, decoder, operationContext),
+                        operationContext);
+            }
+            throw e;
+        }
+    }
+
+    @Override
+    public <T> void sendAndReceiveAsync(final CommandMessage message, final Decoder<T> decoder,
+                                        final OperationContext operationContext,
+                                        final SingleResultCallback<T> callback) {
+
+        AsyncSupplier<T> sendAndReceiveAsyncInternal = c -> sendAndReceiveAsyncInternal(
+                message, decoder, operationContext, c);
+        beginAsync()
+                .thenSupply(sendAndReceiveAsyncInternal::getAsync)
+                .onErrorIf(this::reauthenticationIsTriggered, (t, c) -> {
+                    reauthenticateAndRetryAsync(sendAndReceiveAsyncInternal, operationContext, c);
+                }).finish(callback);
```

Comment:
These async API blocks should be formatted in the original way:

```suggestion
        beginAsync().<T>thenSupply(c -> {
            sendAndReceiveAsyncInternal.getAsync(c);
        }).onErrorIf(e -> reauthenticationIsTriggered(e), (t, c) -> {
            reauthenticateAndRetryAsync(sendAndReceiveAsyncInternal, operationContext, c);
        }).finish(callback);
```
It is better to put all of the boilerplate on the left indentation level, making it easy to ignore, and all of the "real code" at the same indentation level, where it can easily be scanned and compared against its sync counterpart. Even in trivial cases, this is preferable for consistency.

Though I see no issue with `this::reauthenticationIsTriggered`, I have left the original form - let's avoid doing any refactoring during a merge, unless necessary.